modules/puppet/templates/govuk_puppet

```diff
@@ -8,11 +8,222 @@ fi
+	echo "	--list-locks : show all current locks "
+	echo "	--remove-lock : remove one lock, takes numeric identifier as parameter "
+	echo "	--disable : add a lock "
+	echo "	--enable : remove all locks "
+
+	exit 1
+
+}
+
+PUPPET_STATE_DIR=`puppet config print statedir`
+GOVUK_STATE_DIR="$PUPPET_STATE_DIR/govuk"
+
+#make it if it isn't
+[ -d $GOVUK_STATE_DIR ] || mkdir "$GOVUK_STATE_DIR"
+
+#change working directory or die
+cd "$GOVUK_STATE_DIR" || { echo "Unable to access locks directory: '$GOVUK_STATE_DIR'" 1>&2 ; exit 1; }
+
+#echo "State directory:" "$GOVUK_STATE_DIR"
+#echo
+
+case "$1" in
+
+--disable)
+
+	#require a reason, determine a name if interactive
+	#append a datetime stamp to assist in locating old / forgotten locks
+	#write the lockfile containing the reason and name, then 
+	#do puppet agent --disable
+
+	LOCK_DATE=`date +%Y-%m-%d`	
+
+	if [ "$TERM" == "dumb" ] #Test for interactive shell
+	then
+		IAM="system"
+	else
+		#Capture initial login user
+		IAM=`who -am | awk '{ print $1 }'`
+	fi
+	
+	REASON=${2:-} #Take care of the annoying -u with parameter expansion
+
+	if [ -z "$REASON" ] #Check if a reason has been passed
+	then 
+		#It wasn't
+		echo "Reason required as second parameter (quote-wrap); exiting .."
+		exit 1;
+	else
+		REASON="$2"
+	fi
+	
+   	#Create the lock
+		MYID=`date +%s`
+		LOCKFILE=puppet_disable_lock.$IAM.$LOCK_DATE.$MYID.lock
+		echo "Creating $LOCKFILE .."
+		touch $LOCKFILE || { echo "Unable to create lock  '$LOCKFILE'" 1>&2 ; exit 1; }
+		echo "$REASON - $MYID" >> $LOCKFILE
+		do_puppet #Disable puppet agent - helpfully the parameter required is the same
+	
+	exit 0;
+	;;
+
+--list-locks)
+
+	listlocks;
+	exit 0;
+	;;
+		
+--remove-lock)
+
+	#remove lock by identifier
+
+	MYID=${2:-}
+
+	if [ -z "$MYID" ] #If no identifier is passed, exit.
+	then
+		echo "Identifier required; exiting.."
+		exit 1;
+	fi
+
+	if ls *.$MYID.lock 1> /dev/null 2>&1; then #if the lock exists remove it
+
+		for filename in *.$MYID.lock
+                	do
+				rm $filename && echo "Removed lock: $filename";
+                	done
+
+		#Now check if it any locks remain. If not, re-enable puppet agent.
+
+		if ls *.lock 1> /dev/null 2>&1; then #there are more locks
+
+			exit 0;
+
+		else #there are no locks left, re-enable
+
+			# Allow return code 2 in case -v/--detailed-exitcodes is passed.
+        		RBENV_VERSION=1.9.3 /usr/local/bin/govuk_setenv default puppet agent --onetime --no-daemonize --enable || [ $? -eq 2 ]
+
+        		TIME_TOOK=$((`date +%s%3N`-START_TIME))
+        		echo "<%= @fqdn_metrics %>.puppet.run_duration:${TIME_TOOK}|ms" | nc -w 1 -u localhost 8125
+
+        		# We should only reach here if a last_run_summary was written.
+        		/usr/local/bin/puppet_passive_check_update >/dev/null
+        		exit 0;
+
+		fi
+	
+		exit 0;
+	
+	else
+		echo "No lock with identifier: $MYID found. Exiting.."
+		exit 1;
+	fi
+
+	;;
+
+--enable)
+
+	#TODO - interactive vs non-interactive behaviour
```

Comment:
I actually don't think we need interactive behaviour. We just need to be able to pass the required arguments, and throw an error if people don't pass them.